File: backend/src/services/geocoding.ts
import type { Place } from "../types/index.js";
import { config } from "../config.js";

interface NominatimResult {
  display_name?: string;
  lat?: string;
  lon?: string;
  name?: string;
}

/**
 * Geocode a free-text query via OpenStreetMap Nominatim.
 * Public API: keep requests polite (rate limiting is applied at the route layer).
 */
export async function geocode(query: string): Promise<Place[]> {
  const url = new URL("/search", config.nominatimUrl);
  url.searchParams.set("format", "jsonv2");
  url.searchParams.set("q", query);
  url.searchParams.set("limit", "5");
  url.searchParams.set("addressdetails", "1");

  const res = await fetch(url, {
    signal: AbortSignal.timeout(8000),
    headers: {
      "User-Agent": "AccessiPath/1.0 (hackathon accessibility routing; contact: team@accessipath.dev)",
      Accept: "application/json",
    },
  });
  if (!res.ok) throw new Error(`Nominatim responded with HTTP ${res.status}`);
  const data = (await res.json()) as NominatimResult[];
  return data
    .map((r): Place | null => {
      const lat = Number(r.lat);
      const lon = Number(r.lon);
      if (!Number.isFinite(lat) || !Number.isFinite(lon)) return null;
      return {
        id: `nom-${encodeURIComponent(r.display_name ?? "unknown").slice(0, 60)}`,
        label: r.name || (r.display_name ?? "Unknown place").split(",")[0]!,
        description: r.display_name ?? "",
        latitude: lat,
        longitude: lon,
        source: "nominatim",
      };
    })
    .filter((p): p is Place => p !== null);
}

File: backend/app/services/geocoding.py
"""Geocoding via OpenStreetMap Nominatim (port of the Node geocoding service)."""

import math
import urllib.parse

import httpx

from ..config import settings
from ..schemas import Place


async def geocode(query: str) -> list[Place]:
    url = f"{settings.nominatim_url.rstrip('/')}/search"
    params = {
        "format": "jsonv2",
        "q": query,
        "limit": "5",
        "addressdetails": "1",
    }
    async with httpx.AsyncClient(timeout=8.0) as client:
        response = await client.get(
            url,
            params=params,
            headers={
                "User-Agent": (
                    "AccessiPath/1.0 (hackathon accessibility routing; "
                    "contact: team@accessipath.dev)"
                ),
                "Accept": "application/json",
            },
        )
    if response.status_code >= 300:
        raise RuntimeError(f"Nominatim responded with HTTP {response.status_code}")

    results: list[Place] = []
    for item in response.json():
        try:
            lat = float(item.get("lat"))
            lon = float(item.get("lon"))
        except (TypeError, ValueError):
            continue
        if not (math.isfinite(lat) and math.isfinite(lon)):
            continue
        display_name = item.get("display_name") or ""
        name = item.get("name") or display_name.split(",")[0] or "Unknown place"
        results.append(
            Place(
                id=f"nom-{urllib.parse.quote(display_name)[:60]}",
                label=name,
                description=display_name,
                latitude=lat,
                longitude=lon,
                source="nominatim",
            )
        )
    return results